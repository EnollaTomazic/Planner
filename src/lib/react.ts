import * as React from "react";

export const hasTextContent = (node: React.ReactNode): boolean => {
  if (node === null || node === undefined) return false;
  if (typeof node === "boolean") return false;
  if (typeof node === "string") return node.trim().length > 0;
  if (typeof node === "number") return true;
  if (Array.isArray(node)) return node.some((item) => hasTextContent(item));
  if (React.isValidElement<{ children?: React.ReactNode }>(node)) {
    return hasTextContent(node.props.children ?? null);
  }
  return false;
};

export function observeMediaQuery(
  query: string,
  onChange: (matches: boolean, mediaQuery: MediaQueryList) => void,
): (() => void) | undefined {
  if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
    return undefined;
  }

  let mediaQuery: MediaQueryList;

  try {
    mediaQuery = window.matchMedia(query);
  } catch {
    return undefined;
  }

  const handleChange = (event: MediaQueryListEvent): void => {
    onChange(event.matches, mediaQuery);
  };

  const handleLegacyChange = (): void => {
    onChange(mediaQuery.matches, mediaQuery);
  };

  onChange(mediaQuery.matches, mediaQuery);

  if (typeof mediaQuery.addEventListener === "function") {
    mediaQuery.addEventListener("change", handleChange);
    return () => {
      mediaQuery.removeEventListener("change", handleChange);
    };
  }

  if (typeof mediaQuery.addListener === "function") {
    mediaQuery.addListener(handleLegacyChange);
    return () => {
      mediaQuery.removeListener(handleLegacyChange);
    };
  }

  return undefined;
}

export function useMatchMedia(
  query: string,
  getServerSnapshot?: () => boolean,
): boolean {
  const getMatches = React.useCallback(() => {
    if (typeof window === "undefined") {
      return typeof getServerSnapshot === "function" ? getServerSnapshot() : false;
    }

    if (typeof window.matchMedia !== "function") {
      return false;
    }

    try {
      return window.matchMedia(query).matches;
    } catch {
      return false;
    }
  }, [query, getServerSnapshot]);

  const [matches, setMatches] = React.useState<boolean>(getMatches);

  React.useEffect(() => {
    return observeMediaQuery(query, (nextMatches) => {
      setMatches(nextMatches);
    });
  }, [query]);

  return matches;
}
