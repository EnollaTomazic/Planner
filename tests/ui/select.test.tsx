import * as React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { describe, it, expect, afterEach } from "vitest";
import Select from "../../src/components/ui/Select";

afterEach(cleanup);

describe("Select", () => {
  it("renders default state", () => {
    const { container } = render(
      <Select
        variant="native"
        aria-label="test"
        items={[
          { value: "", label: "Choose…" },
          { value: "a", label: "A" },
        ]}
        value=""
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders focus state", () => {
    const { container, getByRole } = render(
      <Select
        variant="native"
        aria-label="test"
        items={[
          { value: "", label: "Choose…" },
          { value: "a", label: "A" },
        ]}
        value=""
      />,
    );
    fireEvent.focus(getByRole("combobox"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders error state", () => {
    const { container } = render(
      <Select
        variant="native"
        aria-label="test"
        errorText="Error"
        items={[{ value: "", label: "Choose…" }]}
        value=""
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders success state", () => {
    const { container } = render(
      <Select
        variant="native"
        aria-label="test"
        success
        items={[{ value: "", label: "Choose…" }]}
        value=""
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders disabled state", () => {
    const { container } = render(
      <Select
        variant="native"
        aria-label="test"
        disabled
        items={[{ value: "", label: "Choose…" }]}
        value=""
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has no outline when focused", () => {
    const { getByRole } = render(
      <Select
        variant="native"
        aria-label="outline"
        items={[{ value: "", label: "Choose…" }]}
        value=""
      />,
    );
    const select = getByRole("combobox");
    fireEvent.focus(select);
    const style = getComputedStyle(select);
    expect(style.outlineStyle === "none" || style.outlineStyle === "").toBe(
      true,
    );
    expect(style.outlineWidth === "0px" || style.outlineWidth === "").toBe(
      true,
    );
  });
});
