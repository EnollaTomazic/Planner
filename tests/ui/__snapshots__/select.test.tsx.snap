// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > renders default state 1`] = `
<div
  class="space-y-1"
>
  <div
    class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,hsl(var(--accent-2)/0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] group jitter hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)]"
    style="--theme-ring: hsl(var(--ring));"
  >
    <select
      aria-label="test"
      class="flex-1 h-11 px-[var(--space-14)] pr-[var(--space-36)] text-sm bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))/0.8] caret-[hsl(var(--accent))] appearance-none disabled:cursor-not-allowed focus:outline-none focus-visible:outline-none"
      id=":r0:"
      name="test"
    >
      <option
        value=""
      >
        Choose…
      </option>
      <option
        value="a"
      >
        A
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down pointer-events-none absolute right-[var(--space-14)] h-4 w-4 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))]"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </div>
</div>
`;

exports[`Select > renders disabled state 1`] = `
<div
  class="space-y-1"
>
  <div
    class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-[--theme-ring] focus-within:ring-offset-0 before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,hsl(var(--accent-2)/0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] opacity-60 pointer-events-none group jitter hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)] cursor-not-allowed focus-within:ring-0 focus-within:shadow-none"
    style="--theme-ring: hsl(var(--ring));"
  >
    <select
      aria-label="test"
      class="flex-1 h-11 px-[var(--space-14)] pr-[var(--space-36)] text-sm bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))/0.8] caret-[hsl(var(--accent))] appearance-none disabled:cursor-not-allowed focus:outline-none focus-visible:outline-none"
      disabled=""
      id=":r4:"
      name="test"
    >
      <option
        value=""
      >
        Choose…
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down pointer-events-none absolute right-[var(--space-14)] h-4 w-4 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))]"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </div>
</div>
`;

exports[`Select > renders error state 1`] = `
<div
  class="space-y-1"
>
  <div
    class="relative inline-flex w-full items-center rounded-xl border bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,hsl(var(--accent-2)/0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] border-[hsl(var(--danger)/0.6)] focus-within:ring-[hsl(var(--danger)/0.35)] group jitter hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)]"
    style="--theme-ring: hsl(var(--ring));"
  >
    <select
      aria-describedby=":r2:-error"
      aria-invalid="true"
      aria-label="test"
      class="flex-1 h-11 px-[var(--space-14)] pr-[var(--space-36)] text-sm bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))/0.8] caret-[hsl(var(--accent))] appearance-none disabled:cursor-not-allowed focus:outline-none focus-visible:outline-none"
      id=":r2:"
      name="test"
    >
      <option
        value=""
      >
        Choose…
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down pointer-events-none absolute right-[var(--space-14)] h-4 w-4 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))]"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </div>
  <p
    class="text-xs mt-1 line-clamp-2 text-danger"
    id=":r2:-error"
  >
    Error
  </p>
</div>
`;

exports[`Select > renders focus state 1`] = `
<div
  class="space-y-1"
>
  <div
    class="relative inline-flex w-full items-center rounded-xl border border-[hsl(var(--border)/0.28)] bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-[--theme-ring] focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,hsl(var(--accent-2)/0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] group jitter hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)]"
    style="--theme-ring: hsl(var(--ring));"
  >
    <select
      aria-label="test"
      class="flex-1 h-11 px-[var(--space-14)] pr-[var(--space-36)] text-sm bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))/0.8] caret-[hsl(var(--accent))] appearance-none disabled:cursor-not-allowed focus:outline-none focus-visible:outline-none"
      id=":r1:"
      name="test"
    >
      <option
        value=""
      >
        Choose…
      </option>
      <option
        value="a"
      >
        A
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down pointer-events-none absolute right-[var(--space-14)] h-4 w-4 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))]"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </div>
</div>
`;

exports[`Select > renders success state 1`] = `
<div
  class="space-y-1"
>
  <div
    class="relative inline-flex w-full items-center rounded-xl border bg-[hsl(var(--card)/0.6)] backdrop-blur-[2px] shadow-[0_0_0_1px_hsl(var(--border)/0.12)] transition-[box-shadow,transform] duration-150 ease-out hover:border-[hsl(var(--border)/0.38)] focus-within:outline-none focus-within:ring-2 focus-within:ring-offset-0 focus-within:shadow-[0_0_24px_hsl(var(--accent)/0.14)] before:pointer-events-none before:absolute before:inset-0 before:rounded-[inherit] before:opacity-[0.05] before:bg-[repeating-linear-gradient(0deg,hsl(var(--accent-2)/0.4)_0_1px,transparent_1px_3px)] after:pointer-events-none after:absolute after:inset-0 after:rounded-[inherit] after:opacity-[0.04] after:bg-[url('/noise.svg')] group jitter hover:shadow-[0_0_0_1px_hsl(var(--border)/0.2)] border-[--theme-ring] focus-within:ring-[--theme-ring]"
    style="--theme-ring: hsl(var(--ring));"
  >
    <select
      aria-describedby=":r3:-success"
      aria-label="test"
      class="flex-1 h-11 px-[var(--space-14)] pr-[var(--space-36)] text-sm bg-transparent text-[hsl(var(--foreground))] placeholder:text-[hsl(var(--muted-foreground))/0.8] caret-[hsl(var(--accent))] appearance-none disabled:cursor-not-allowed focus:outline-none focus-visible:outline-none"
      id=":r3:"
      name="test"
    >
      <option
        value=""
      >
        Choose…
      </option>
    </select>
    <svg
      aria-hidden="true"
      class="lucide lucide-chevron-down pointer-events-none absolute right-[var(--space-14)] h-4 w-4 text-[hsl(var(--muted-foreground))] group-focus-within:text-[hsl(var(--accent))]"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </div>
  <p
    aria-live="polite"
    class="sr-only"
    id=":r3:-success"
    role="status"
  >
    Selection saved
  </p>
</div>
`;
